test/new-e2e/utils/e2e/client/docker.go

```diff
@@ -0,0 +1,109 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/DataDog/test-infra-definitions/components/datadog/agent/docker"
+	"github.com/docker/cli/cli/connhelper"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/stretchr/testify/require"
+)
+
+var _ clientService[docker.ClientData] = (*Docker)(nil)
+
+// A docker client that is connected to an [docker.Deamon].
+//
+// [docker.Deamon]: https://pkg.go.dev/github.com/DataDog/test-infra-definitions@main/components/datadog/agent/docker#Deamon
+type Docker struct {
+	*UpResultDeserializer[docker.ClientData]
+	t                  *testing.T
+	client             *client.Client
+	agentContainerName string
+}
+
+// Create a new instance of Docker
+func NewDocker(daemon *docker.Daemon) *Docker {
+	dockerInstance := &Docker{
+		agentContainerName: daemon.GetAgentContainerName(),
+	}
+	dockerInstance.UpResultDeserializer = NewUpResultDeserializer[docker.ClientData](daemon, dockerInstance)
+	return dockerInstance
+}
+
+//lint:ignore U1000 Ignore unused function as this function is call using reflection
+func (docker *Docker) initService(t *testing.T, data *docker.ClientData) error {
+	docker.t = t
+
+	deamonURL := fmt.Sprintf("ssh://%v@%v:22", data.Connection.User, data.Connection.Host)
+	helper, err := connhelper.GetConnectionHelperWithSSHOpts(deamonURL, []string{"-o", "StrictHostKeyChecking no"})
+
+	if err != nil {
+		return fmt.Errorf("cannot connect to docker %v: %v", deamonURL, err)
+	}
+
+	opts := []client.Opt{
+		client.WithDialContext(helper.Dialer),
+		client.WithAPIVersionNegotiation(),
+	}
+
+	docker.client, err = client.NewClientWithOpts(opts...)
+	return err
+}
+
+func (docker *Docker) ExecuteCommand(containerName string, commands ...string) string {
+	output, errOutput, err := docker.ExecuteCommandWithErr(containerName, commands...)
+	if len(errOutput) != 0 {
+		output += " " + errOutput
+	}
+
+	require.NoErrorf(docker.t, err, "%v", output)
+	return output
+}
+
+func (docker *Docker) ExecuteCommandWithErr(containerName string, commands ...string) (string, string, error) {
+	context := context.Background()
+	execConfig := types.ExecConfig{Cmd: commands, AttachStderr: true, AttachStdout: true}
+	execCreateResp, err := docker.client.ContainerExecCreate(context, containerName, execConfig)
+	require.NoError(docker.t, err)
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
Why not returning error instead ? This will panic the test in case of error

Reply:
This function executes a command on Docker and so there are two kinds of error:
1. The error related to the internal logic to execute a command on Docker. This is part of the framework.
2. The error returned by user's command.

Here, the error is not returned because it is part of the framework (`1.`) and the user cannot really do something useful with this error.